DataDog/integrations-core
# Add VoltDB integration

### What does this PR do?

Add an integration for VoltDB.

### Motivation


### Additional Notes


TODO:

* [x] Test against VoltDB 10.0

NOTE: logs support will be added as a follow-up PR.

### Review checklist (to be filled by reviewers)


### QA Notes

* Context: we're releasing this integration to a customer who's running Agent 7.22 and VoltDB 8.4. So we want to make sure 8.4 in particular is smooth as butter.
* Recommended envs to test: `py27-8.4`, `py38-8.4-tls`, `py38-10.0`.
* Consider re-reviewing the PR, see if there's anything that looks off.
* Start the env, verify metrics look good. FWIW, see the initial integration requirements below, feel free to double-check they seem to be well-covered by the integration.
* Double-check both tests and E2E work.
* In the `-tls` env, double check you can access the API through `curl` (to make sure the TLS setup is correct):

```bash
curl --cacert $(pwd)/voltdb/tests/tlsoutput/client.pem https://localhost:8443/api/1.0 -d 'Procedure=@SystemInformation' -d 'User=doggo' -d 'Password=doggopass'
```

Initial integration requirements:

```
* Provide per-node information for CPU usage → CPU component
* Provide per-node information for memory usage → MEMORY component
* Enhanced information about snapshots: Add information about the size and duration of each completed snapshot → SNAPSHOTSTATUS component, SIZE DURATION
* Enhanced information about commandlogs → COMMANDLOG component
    * Show the FSYNC information per-node → FSYNC_INTERVAL column, tagged with HOST_ID / HOSTNAME
    * Add outstanding transactions to be command logged per-node → OUTSTANDING_TXNS column, tagged with HOST_ID / HOSTNAME
* Detailed Procedure Statistics →  PROCEDURE component
* Stats for cluster → LATENCY component provides p50, p95, p99, p99.9, p99.99, p99.999 (sic) and max for latency, and monotonic count of transactions
* Procedure calls/second for each procedure → PROCEDURE component, INVOCATIONS column (monotonic count, which they can view as a rate in UI)
* Invocation count and avg/p95/p99 execution time for individual statements. → PROCEDUREDETAIL component, INVOCATIONS (monotonic count). We can only provide MIN/MAX/AVG execution times (p95 and p99 not available)
* JVM GC → GC component
    * Stats per-node → Tagged by HOST_ID / HOSTNAME
    * Average time and counts in oldgen and newgen GC →  NEWGEN_*, OLDGEN_*
* Network &  Disk IO → IOSTATS (bytes/messages read/written). NOTE: didn’t find anything about disk I/O, so we don't have specific metrics for that.
* Table metrics: → TABLE component
    * Rows per table → TUPLE_COUNT
    * Ram usage per table → TUPLE_ALLOCATED, TUPLE_DATA_MEMORY, STRING_DATA​_MEMORY
    * Index count, Memory used → INDEX component, tagged by INDEX_NAME, TABLE_NAME. They can get index count per table by aggregating in UI. Memory used by index: MEMORY_ESTIMATE
```
